Feature/data models

IvanMukticc opened the pull request:
Adding DataModel,
Making API call with async function and URLSession,
Adding KingfisherSwiftUI

IvanMukticcc committed:
Adding data model,
Adding API manager,
Making API call.

IvanMukticcc committed:
Getting API with async function,
Cleaning ApiManager,
Adding new package "KingfisherSwiftUI".

horvatrobert commented on Sources/ApiManager.swift:
```diff
@@ -0,0 +1,50 @@
+    func getRepositories(for search: String) async throws -> [Repository] {
+        guard let urlRequest = getUrlRequest(searchTerm: search)
+        else {
+            return []
```
throw error here

horvatrobert commented on Sources/ApiManager.swift:
```diff
@@ -0,0 +1,50 @@
+        }
+
+        let decoded = try decoder.decode(Root.self, from: data)
+        print(decoded.items?.count ?? 0)
```
remove any unnecessary prints from the codebass

horvatrobert commented on Sources/ContentView.swift:
```diff
@@ -1,11 +1,29 @@
+            VStack {
+                List {
+                    ForEach(repositories) { repository in
+                        Text(repository.name ?? "Pero")
```
Avoid using such default values. Instead you can use "n/a"

horvatrobert commented on Sources/ContentView.swift:
```diff
@@ -1,11 +1,29 @@
     var body: some View {
         NavigationView {
-            VStack {}
+            VStack {
```
VStack wrapper is not necessary here, List itself will do the job :)

horvatrobert commented on Sources/ContentView.swift:
```diff
@@ -1,11 +1,29 @@
+                    repositories = try await apiManager
+                        .getRepositories(for: searchTerm)
+                } catch {
+                    print(error.localizedDescription)
```
You can implement alert with the error message caught here.

horvatrobert commented on Sources/DataModel.swift:
```diff
@@ -0,0 +1,39 @@
+    let language: String?
+    let owner: Owner?
+
+    static let example = Repository(
```
instead of example you can rename it to mock since this is mock data.

horvatrobert commented on project.yml:
```diff
@@ -24,7 +27,9 @@ targetTemplates:
         scheme: {}
-        dependencies: []
+        dependencies:
+            - package: KingfisherSwiftUI
```
I'm not 100% sure if this is correct but you will see when you get to fetching images from url-s.

horvatrobert commented on project.yml:
```diff
@@ -24,7 +27,9 @@ targetTemplates:
             - Config/${enviroment}
         scheme: {}
-        dependencies: []
+        dependencies:
```
Instead of kingfished use AsyncImage which is available in SwiftUI from iOS15.

https://www.hackingwithswift.com/quick-start/swiftui/how-to-load-a-remote-image-from-a-url

IvanMukticcc committed:
Throwing error,
Removing unnecessary prints and dependencies,
Implementing alert.

IvanMukticcc committed:
Adding error,
Adding .searchable,
Adding delay on searchTerm,
Improving alert

horvatrobert commented on Sources/ContentView.swift:
```diff
@@ -5,25 +5,32 @@ struct ContentView: View {
     @State
-    private var searchTerm: String = "php"
+    private var searchTerm: String = ""
+    @State
```
Instead of isError rename the variable somehow different. Variable needs to be named to be clear and self-explainatory.
Good example would be errorPresented or alertPresented or combination of those two errorAlertPresented

horvatrobert commented on Sources/ContentView.swift:
```diff
@@ -5,25 +5,32 @@ struct ContentView: View {
+    @State
+    private var isError = false
+    @State
+    private var message = ""
```
Same here. Message could mean anything. Rename it to contain more detailed description.
errorMessage would be perfect in this particular case.

horvatrobert commented on Sources/ContentView.swift:
```diff
@@ -5,25 +5,32 @@ struct ContentView: View {
-                    }
+            List {
+                ForEach(repositories) { repository in
+                    Text(repository.name ?? "N/A")
```
Just a personal preference, I would put lowercased "n/a" instead of uppercased "N/A". Also I think lowercased is more common in codebases.

_n/a is a common abbreviation in tables and lists for the phrase not applicable, not available, not assessed, or no answer._

horvatrobert commented on Sources/ApiManager.swift:
```diff
@@ -3,11 +3,11 @@ import Foundation
         case invalidURL
-        case dataNotFound
-        case decodingError
+        case dataNotFound(response: Int?)
```
why adding response code to the error which implies 404? (not found)
Maybe just log the status code (print("An error occured with status: \(response)")

horvatrobert replied:
Or rename the error from dataNotFound to generic(withResponse response: Int)

karlovcec commented on Sources/ApiManager.swift:
```diff
@@ -0,0 +1,49 @@
+    enum ApiError: Error, LocalizedError {
+        case invalidURL
+        case dataNotFound(response: Int?)
+        case badURLRequest
```
The whole idea of conforming to LocalizedError protocol is to have a string representation of each error. Eventually, you will want to handle these errors in UI layer.  The errorDescription property will come in handy :)
eg:
```swift
enum ApiError: Error, LocalizedError {
  case invalidURL

  var errorDescription: String? {
    switch self {
      case .invalidURL:
        return "Invalid URL!!"
      }
    }
}
```

horvatrobert commented on Sources/ContentView.swift:
```diff
@@ -11,23 +11,29 @@ struct ContentView: View {
-                        "Could no load data. \(error.localizedDescription): Data not found."
-                    isError = true
+            .searchable(text: $searchTerm, prompt: "Please enter search term")
+            .onChange(of: searchTerm, perform: { _ in
```
here instead of _ you can use newSearchTerm and send this to the api call. 
onChange returns new value in this closure and you can do with it what needs to be done.
This might be related to the issue we talked about 10 mins ago.

IvanMukticcc committed:
Adding string representation of errors,
Adding timer to delay triggering request,
Renaming ContentView (for readability).

horvatrobert reviewed (changes requested):
I have no additional comments on the current state of the codebase.
Only logical step atm would be to extract all of the properties to dedicated RepositoryListViewModel which would be marked as @MainActor and every logic which is currently happening in the view itself, and fetching the data will happen there. 

🤓 Hint:

RepositoryListViewModel.swift
```swift
extension RepositoryListView {
    @MainActor
    class ViewModel: ObservableObject {

    }
}
```

RepositoryListView.swift
Here instead having all of the properties you currently have will only have

```swift
@StateObject
var viewModel: ViewModel
```

IvanMukticcc committed:
Adding view model for RepositoryListView.
